sdk/iot/core/tests/cmocka/az_iot_telemetry_tests.c

```diff
@@ -0,0 +1,301 @@
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_MODULES_MID
+                                "" TEST_MODULE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS;
+static const char g_test_correct_topic_with_options_user_agent_with_params[]
+    = TEST_CORRECT_TOPIC_PREFIX "" TEST_DEVICE_ID "" TEST_CORRECT_TOPIC_SUFFIX "?" TEST_PARAMS
+                                "&" TEST_USER_AGENT;
+
+static az_iot_hub_client g_test_valid_client_no_options
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_EMPTY_OPTIONS } };
+static const az_iot_hub_client g_test_valid_client_with_options_both
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_BOTH } };
+static const az_iot_hub_client g_test_valid_client_with_options_module_id
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_MODULE_ID } };
+static const az_iot_hub_client g_test_valid_client_with_options_user_agent
+    = { ._internal = { .iot_hub_hostname = AZ_SPAN_LITERAL_FROM_STR(TEST_FQDN),
+                       .device_id = AZ_SPAN_LITERAL_FROM_STR(TEST_DEVICE_ID),
+                       .options = TEST_VALID_OPTIONS_USER_AGENT } };
+
+void az_iot_hub_client_telemetry_publish_topic_get_NULL_client_fails(void** state)
+{
+  (void)state;
+
+  uint8_t mqtt_topic_buf[TEST_MQTT_SPAN_BUFFER_SIZE];
+
+  az_span mqtt_topic = AZ_SPAN_LITERAL_FROM_BUFFER(mqtt_topic_buf);
+
+  assert_true(
+      az_iot_hub_client_telemetry_publish_topic_get(NULL, NULL, mqtt_topic, &mqtt_topic)
+      == AZ_ERROR_ARG);
+}
+
+void az_iot_hub_client_telemetry_publish_topic_get_NULL_mqtt_topic_fails(void** state)
+{
+  (void)state;
+  az_span null_mqtt_topic = TEST_NULL_PTR_SPAN;
+
+  assert_true(
+      az_iot_hub_client_telemetry_publish_topic_get(
+          &g_test_valid_client_no_options, NULL, null_mqtt_topic, &null_mqtt_topic)
+      == AZ_ERROR_ARG);
+}
+
+void az_iot_hub_client_telemetry_publish_topic_get_NULL_out_mqtt_topic_fails(void** state)
+{
+  (void)state;
+  uint8_t mqtt_topic_buf[TEST_MQTT_SPAN_BUFFER_SIZE];
+
+  az_span mqtt_topic = AZ_SPAN_LITERAL_FROM_BUFFER(mqtt_topic_buf);
+
+  assert_true(
+      az_iot_hub_client_telemetry_publish_topic_get(
+          &g_test_valid_client_no_options, NULL, mqtt_topic, NULL)
+      == AZ_ERROR_ARG);
+}
+
+void az_iot_hub_client_telemetry_publish_topic_get_no_options_no_params_succeed(void** state)
+{
+  (void)state;
+
+  uint8_t mqtt_topic_buf[sizeof(g_test_correct_topic_no_options_no_params)];
+
+  az_span mqtt_topic = AZ_SPAN_LITERAL_FROM_BUFFER(mqtt_topic_buf);
```

Comment:
>AZ_SPAN_LITERAL_FROM_BUFFER [](start = 23, length = 27)

I do not understand this macro

Does this create span from buffer? If yes, then what is Literal in this macro means?

Reply:
You can see the reasoning for "literal" here in a PR I have:

https://github.com/Azure/azure-sdk-for-c/pull/419

In this case I think I can actual do without the literal part. I'll double check how I use it in tests